source/server/admin/stats_render.cc

```diff
@@ -0,0 +1,304 @@
+    } else {
+      ASSERT(histogram_buckets_mode_ != Utility::HistogramBucketsMode::NoBuckets);
+      histograms_obj_container_fields["histograms"] = ValueUtil::listValue(histogram_array_);
+    }
+    auto str = MessageUtil::getJsonStringFromMessageOrDie(
+        ValueUtil::structValue(histograms_obj_container_), false /* pretty */, true);
+    addStatAsRenderedJson(response, str);
+  }
+  response.add("]}");
+}
+
+// Collects a scalar metric (text-readout, counter, or gauge) into an array of
+// stats, so they can all be serialized in one shot when a threshold is
+// reached. Serializing each one individually results in much worse
+// performance (see stats_handler_speed_test.cc).
+template <class Value>
+void StatsJsonRender::addScalar(Buffer::Instance& response, const std::string& name,
+                                const Value& value) {
+  ProtobufWkt::Struct stat_obj;
+  ProtoMap& stat_obj_fields = *stat_obj.mutable_fields();
+  stat_obj_fields["name"] = ValueUtil::stringValue(name);
+  stat_obj_fields["value"] = value;
+  addJson(response, ValueUtil::structValue(stat_obj));
+}
+
+// Adds a JSON stat to our buffer, flushing to response every JsonStatsFlushCount stats.
+void StatsJsonRender::addJson(Buffer::Instance& response, ProtobufWkt::Value json) {
+  stats_array_.push_back(json);
+
+  // We build up stats_array to a certain size so we can amortize the overhead
+  // of entering into the JSON serialization infrastructure. If we set the
+  // threshold too high we buffer too much memory, likely impacting processor
+  // cache. The optimum threshold found after a few experiments on a local
+  // host appears to be between 50 and 100.
+  if (stats_array_.size() == JsonStatsFlushCount) {
+    flushStats(response);
+  }
+}
+
+// Flushes all stats that were buffered in addJson() above.
+void StatsJsonRender::flushStats(Buffer::Instance& response) {
+  ASSERT(!stats_array_.empty());
+  const std::string json_array = MessageUtil::getJsonStringFromMessageOrDie(
+      ValueUtil::listValue(stats_array_), false /* pretty */, true);
+  stats_array_.clear();
+
+  // We are going to wind up with multiple flushes which have to serialize as
+  // a single array, rather than a concatenation of multiple arrays, so we add
+  // those in the constructor and finalize() method, strip off the "[" and "]"
+  // from each buffered serialization.
+  ASSERT(json_array.size() >= 2);
+  ASSERT(json_array[0] == '[');
+  ASSERT(json_array[json_array.size() - 1] == ']');
+  addStatAsRenderedJson(response, absl::string_view(json_array).substr(1, json_array.size() - 2));
+}
+
+// Adds a json fragment of scalar stats to the response buffer, including a
+// "," delimiter if this is not the first fragment.
+void StatsJsonRender::addStatAsRenderedJson(Buffer::Instance& response, absl::string_view json) {
+  if (first_) {
+    response.add(json);
+    first_ = false;
+  } else {
+    response.addFragments({",", json});
+  }
+}
+
```

Comment:
Yes it can be empty, if a name being serialized contains invalid utf8 (I have discovered while working on #20428). 

I am not sure if that would make invalid JSON (two commas in a row) but I added an ASSERT here and a runtime conditional where it looks like that can occur in the histogram call-site.